### PR TITLE
Introduce request and request error interceptors

### PR DESCRIPTION
**Pull request by TharmiganK**
## Purpose

Introducing Request Interceptors and Request Error Interceptors at Listener Configuration.

> Fixes [`HTTP Interceptor phase 1 #2062`](https://github.com/ballerina-platform/ballerina-standard-library/issues/2062)

> Part of [`Introduce HTTP Filters #692`](https://github.com/ballerina-platform/ballerina-standard-library/issues/692)

## Examples

Example of a request interceptor service definition : 
```ballerina
service class SampleRequestInterceptor {
    *http:RequestInterceptor;

    // Request context, Request and Caller is shared among interceptor services
    resource function <httpVerb> <path>(http:RequestContext ctx, http:Request req, http:Caller caller) 
                                   returns http:NextService|error? {
       // Do Something
       return ctx.next(); // Calling next will execute the next interceptor in the pipeline
    }
}
```

Example of a request error interceptor service definition : 
```ballerina
// Error interceptors are only executed when there is an error occurred/returned from previous request 
// interceptor
service class SampleRequestErrorInterceptor {
    *http:RequestErrorInterceptor;

    // error is passed from previous interceptor
    resource function 'default [string... path](http:RequestContext ctx, http:Request req, 
                         http:Caller caller, error err) returns http:NextService|error? {
       // Do Something
       return ctx.next(); // Calling next will execute the next interceptor in the pipeline
    }
}
```

Sample listener configuration : 
```ballerina
listener http:Listener serverEP = new(port, 
     config = {
        interceptors : [new SampleRequestInterceptor(), new SampleRequestErrorInterceptor()]
    });
```

For more information refer the [`Proposal: HTTP Interceptors`](https://github.com/ballerina-platform/module-ballerina-http/blob/master/docs/proposals/interceptors.md)

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: ballerina-tests/tests/interceptors_test_common.bal [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import ballerina/http;
+
+service class defaultRequestInterceptor {
+    *http:RequestInterceptor;
+
+    resource function 'default [string...path](http:RequestContext ctx, http:Caller caller, http:Request req) returns http:InterceptorReturnValues {
```
Comment (chamil321) [inline ballerina-tests/tests/interceptors_test_common.bal:6]:
```suggestion
    resource function 'default [string... path](http:RequestContext ctx, http:Caller caller, http:Request req) returns http:InterceptorReturnValues {
```

Comment (chamil321) [reply ballerina-tests/tests/interceptors_test_common.bal:6]: Let's put a space in between

File: ballerina/http_interceptors.bal [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+};
+
+# The HTTP request error interceptor service object  
+public type RequestErrorInterceptor distinct service object {
+
+};
+
+# The return values from an interceptor service function  
+public type NextService RequestInterceptor|RequestErrorInterceptor|Service;
```
Comment (shafreenAnfar) [inline ballerina/http_interceptors.bal:28]: I am just wondering if it should be,

```ballerina
public type NextService RequestInterceptor|Service;
```

Comment (shafreenAnfar) [reply ballerina/http_interceptors.bal:28]: As per the current proposal `next` could only control the flow of RequestInterceptor|Service interceptors.

File: ballerina-tests/tests/interceptors_test_common.bal [hunk 3]
```diff
@@ -0,0 +1,176 @@
+import ballerina/http;
```
Comment (chamil321) [inline ballerina-tests/tests/interceptors_test_common.bal:1]: Missing license header

Comment (TharmiganK) [reply ballerina-tests/tests/interceptors_test_common.bal:1]: Will add that

Comment (TharmiganK) [reply ballerina-tests/tests/interceptors_test_common.bal:17]: Fixed in [`c101943`](https://github.com/ballerina-platform/module-ballerina-http/pull/713/commits/c1019434762e8544d2380b9e543ceb9b74121a23)

File: ballerina/http_interceptors.bal (hunk 2 above)
Comment (TharmiganK) [reply ballerina/http_interceptors.bal]: Fixed in [`c101943`](https://github.com/ballerina-platform/module-ballerina-http/pull/713/commits/c1019434762e8544d2380b9e543ceb9b74121a23)

File: ballerina-tests/tests/request_interceptors_basic_test.bal [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    http:Response res = check requestInterceptorReturnsErrorClientEP->get("/");
+    test:assertEquals(res.statusCode, 500);
+    assertTextPayload(check res.getTextPayload(), "Request interceptor returns an error");
+}
+
+final http:Client requestErrorInterceptorClientEP = check new("http://localhost:" + requestErrorInterceptorTestPort.toString());
+
+listener http:Listener requestErrorInterceptorServerEP = new(requestErrorInterceptorTestPort, config = {
+        interceptors : [new DefaultRequestInterceptor(), new RequestInterceptorReturnsError(), new DefaultRequestErrorInterceptor(), new LastRequestInterceptor()]
+    });
```
Comment (shafreenAnfar) [inline ballerina-tests/tests/request_interceptors_basic_test.bal:74]: Formatting issue.

**Comment by shafreenAnfar** [comment]
Is there a test case to validate the functionality of the `RequestContext` inside `http:Service` ?

Eg - 

```ballerina
service / on new http:Listener(9090) {
    resource function get greeting(RequestContext rctx) returns string {
        rctx.set("foo") // should work 
        rctx.get("foo"); // should work 
        rctx.next() // returns an error
        return "Hello, World!";
    }
}
```

File: ballerina/http_service_endpoint.bal [hunk 5]
```diff
@@ -19,7 +19,7 @@ import ballerina/jballerina.java;
 
 # This is used for creating HTTP server endpoints. An HTTP server endpoint is capable of responding to
 # remote callers. The `Listener` is responsible for initializing the endpoint using the provided configurations.
-public isolated class Listener {
+public class Listener {
```
Comment (shafreenAnfar) [inline ballerina/http_service_endpoint.bal:22]: As discussed let's make it isolated.

File: native/src/main/java/io/ballerina/stdlib/http/api/InterceptorResource.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
+package io.ballerina.stdlib.http.api;
```
Comment (shafreenAnfar) [inline native/src/main/java/io/ballerina/stdlib/http/api/InterceptorResource.java:1]: License header.

File: native/src/main/java/io/ballerina/stdlib/http/api/Service.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import io.ballerina.stdlib.http.transport.message.HttpCarbonMessage;
+import io.ballerina.stdlib.http.uri.URITemplate;
+import io.ballerina.stdlib.http.uri.URITemplateException;
+
+import java.util.List;
+
+/**
+ * Interface for HTTP Service classes.
+ */
+public interface Service {
```
Comment (shafreenAnfar) [inline native/src/main/java/io/ballerina/stdlib/http/api/Service.java:31]: Since this is a public interface better to document the methods.

**Comment by TharmiganK** [comment]
> Is there a test case to validate the functionality of the `RequestContext` inside `http:Service` ?
> 
> Eg -
> 
> ```ballerina
> service / on new http:Listener(9090) {
>     resource function get greeting(RequestContext rctx) returns string {
>         rctx.set("foo") // should work 
>         rctx.get("foo"); // should work 
>         rctx.next() // returns an error
>         return "Hello, World!";
>     }
> }
> ```

No, I can add one for that
Regarding `rctx.next() // returns an error` : what error should we return?

File: native/src/main/java/io/ballerina/stdlib/http/api/InterceptorResource.java (hunk 6 above)
Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/http/api/InterceptorResource.java:18]: Fixed in [`7f4cb18`](https://github.com/ballerina-platform/module-ballerina-http/pull/713/commits/7f4cb1838084bfbb629abafe3b04caa96ba1e09c)

File: ballerina-tests/tests/request_interceptors_basic_test.bal (hunk 4 above)
Comment (TharmiganK) [reply ballerina-tests/tests/request_interceptors_basic_test.bal]: Fixed in [`1100634`](https://github.com/ballerina-platform/module-ballerina-http/pull/713/commits/110063402614b7d064c3a645e09751f3a3c772b5)

File: native/src/main/java/io/ballerina/stdlib/http/api/Service.java (hunk 7 above)
Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/http/api/Service.java:31]: Documented the methods in [`46f14ba `](https://github.com/ballerina-platform/module-ballerina-http/pull/713/commits/46f14ba8dee24952e9c784083802d29e0f29ce10)

**Comment by chamil321** [comment]
> > Is there a test case to validate the functionality of the `RequestContext` inside `http:Service` ?
> > Eg -
> > ```ballerina
> > service / on new http:Listener(9090) {
> >     resource function get greeting(RequestContext rctx) returns string {
> >         rctx.set("foo") // should work 
> >         rctx.get("foo"); // should work 
> >         rctx.next() // returns an error
> >         return "Hello, World!";
> >     }
> > }
> > ```
> 
> No, I can add one for that Regarding `rctx.next() // returns an error` : what error should we return?

This is more like an illegal function invocation in the listener. So we can use `ListenerError`

File: ballerina/http_service_endpoint.bal (hunk 5 above)
Comment (TharmiganK) [reply ballerina/http_service_endpoint.bal:22]: Fixed in [`76d3f76`](https://github.com/ballerina-platform/module-ballerina-http/pull/713/commits/76d3f763edb0eaf25424ea2ab216eba4e1184079)

**Review by chamil321 — APPROVED** [review]

**Pull request merged**